src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -357,564 +398,511 @@ public override unsafe string GetString(byte[] bytes, int byteIndex, int byteCou
-
-                    // Set our internal fallback interesting things.
-                    fallbackBuffer.InternalInitialize(charStart, charEnd, encoder, true);
-                }
+                encoder._charsUsed = numElementsConverted; // signal to any interested encoder that we completed the entire operation
+            }
 
-                Debug.Assert(charLeftOver == 0 || char.IsHighSurrogate(charLeftOver),
-                    "[ASCIIEncoding.GetBytes]leftover character should be high surrogate");
+        Finished:
 
-                // Verify that we have no fallbackbuffer, for ASCII its always empty, so just assert
-                Debug.Assert(!encoder._throwOnOverflow || !encoder.InternalHasFallbackBuffer ||
-                    encoder.FallbackBuffer.Remaining == 0,
-                    "[ASCIICodePageEncoding.GetBytes]Expected empty fallback buffer");
-            }
-            else
-            {
-                fallback = this.EncoderFallback as EncoderReplacementFallback;
-            }
+            return numElementsConverted;
 
+        InvokeFallback:
 
-            // See if we do the fast default or slightly slower fallback
-            if (fallback != null && fallback.MaxCharCount == 1)
-            {
-                // Fast version
-                char cReplacement = fallback.DefaultString[0];
+            // Otherwise, will need to invoke the fallback handler.
+            numElementsConverted = GetBytes_Fallback(pChars, charCount, pBytes, byteCount, encoder, numElementsConverted);
+            goto Finished;
+        }
 
-                // Check for replacements in range, otherwise fall back to slow version.
-                if (cReplacement <= (char)0x7f)
-                {
-                    // We should have exactly as many output bytes as input bytes, unless there's a left
-                    // over character, in which case we may need one more.
-                    // If we had a left over character will have to add a ?  (This happens if they had a funky
-                    // fallback last time, but not this time.) (We can't spit any out though
-                    // because with fallback encoder each surrogate is treated as a seperate code point)
-                    if (charLeftOver > 0)
-                    {
-                        // Have to have room
-                        // Throw even if doing no throw version because this is just 1 char,
-                        // so buffer will never be big enough
-                        if (byteCount == 0)
-                            ThrowBytesOverflow(encoder, true);
-
-                        // This'll make sure we still have more room and also make sure our return value is correct.
-                        *(bytes++) = (byte)cReplacement;
-                        byteCount--;                // We used one of the ones we were counting.
-                    }
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Why does it have to be NoInlining?

Reply:
The JIT inlines the method into the caller (since it's so short, I suppose) without this attribute, but this has a deleterious effect on performance. See the code comments in the method for a more detailed description.